### ISSUE TITLE
RFC: API for Cross-Chain Transfers

### ISSUE THREAD

**Issue by muharem** [paritytech/polkadot-sdk#4715]
## Summary:

This document proposes an API for a call that dispatches an XCM asset transfer from chain X, withdrawing the asset on chain Y and depositing it on chain Z. Chains X, Y, and Z can be different, the same, or any combination thereof. We will illustrate this with the example of the treasury's spend call.

## Motivation:

Currently, the treasury spend call located on chain X (e.g., RelayChain) can transfer its assets to any chain Y (e.g., AssetHub), where Y can be equal to X or different. However, the beneficiary of this transfer must be on chain Y. This setup misses the use case where stakeholders want to move treasury assets from the RelayChain account to the AssetHub account. To avoid this limitation, we currently use a root track to dispatch a call as a treasurer's sovereign account and teleport via the xcm pallet to the desired destination (e.g. AssetHub). It seems reasonable to provide a solution that utilizes the treasury referendum tracks and possibly uses a single spend call for this purpose. 
Additionally, in recent PRs concerning the locatable asset type, reviewers have questioned the need for such a type. Before introducing a similar type to solve the problem described above, I want to share more on the reasoning for such types.

## Proposed Solution:

First, let's review existing production APIs that facilitate different type of cross-chain transfers

### X = Y = Z

This is a simple transfer or the treasury’s `spend_local`. It only requires the `from` account, `to` account, and the `asset id`.

### X ≠ Y, Y = Z

Implemented by the treasury pallet setup of the Polkadot RelayChain:

``` rust
// Dispatched on the X chain.
// Implies `source` = `treasury_account` on withdrawal location, Y chain.
fn treasury_spend (
  ...
  asset_kind: LocatableAssetId {
    // Location of the Y chain.
    location: xcm::Location, 
    // Asset id on the Y chain.
    asset_id: xcm::AssetId,
  }
  // `AccountId` on chain Z or some entity's location (pallet instance on some chain)
  // used to derive it's sovereign account on Z chain.
  // Z must be equal to Y.
  beneficiary: xcm::Location,
  ... 
);
```

It is worth mentioning that a beneficiary location presented as the `xcm::Location` type or an asset id presented as the `xcm::AssetId` (which is a wrapped `xcm::Location`) type cannot carry the information about the chain on which these location are meant to be interpreted. A single location can represent a beneficiary (e.g., the Fellowship Treasury pallet instance on Collectives) or an asset id (e.g., GLMR as Moonbeam location) without an information where they are located.

### X = Y, Y ≠ Z

Implemented by the xcm pallet and its teleport call:

``` rust
// Dispatched on the X chain;
fn xcm_teleport (
  // Implies the `source` = `sovereign_of(origin)` on the X chain.
  origin: Origin,
  // Location of the Z chain.
  dest: xcm::Location,
  // Sovereign account of the given location on Z chain.
  beneficiary: xcm::Location,
  // Location of the assets on Y chain, which must be equal to X.
  assets: Vec<xcm::Asset>,
  …
);
```

The last example is what we need for the treasury use case to transfer the treasury assets from the RelayChain to the AssetHub. If we use a similar API, we would need to introduce a new call like `treasury.withdraw_local_spend_remote`, but we can do better.

### X ≠ Y ≠ Z, and all other possible variations

Drawing from the existing examples, we can propose the following solution:

``` rust 
// Dispatched on the X chain.
// Implies `source` = `treasury_account` on withdrawal location, Y chain.
fn treasury_spend (
  ...
  asset_kind: LocatableAssetId {
    // Location of the Y chain.
    location: xcm::Location, 
    // Asset id on the Y chain.
    asset_id: xcm::AssetId,
  }
  beneficiary: LocatableBeneficiary {
    // Location of the Z chain, deposit location.
    location: xcm::Location,
    // Sovereign account of the given location on Z chain.
    account: xcm::Location,
  }
  …
);
```

The API  can accommodate all desired use cases. The present `Pay` trait will have different implementations for different scenarios and can be chained to find the applicable one.

**Comment by acatangiu** [comment]
Why the need for new type `LocatableBeneficiary`?

```diff
-  beneficiary: LocatableBeneficiary {
-    // Location of the Z chain, deposit location.
-    location: xcm::Location,
-    // Sovereign account of the given location on Z chain.
-    account: xcm::Location,
-  }
+  // Beneficiary's location relative to the current context (chain X),
+  // equivalent location on chain Z can be derived through re-anchoring.
+  beneficiary: xcm::Location,
```

-----

`LocatableAssetId` could also be broken down to equivalent components without the need for a new type:

```diff
-  asset_kind: LocatableAssetId {
-    // Location of the Y chain.
-    location: xcm::Location, 
-    // Asset id on the Y chain.
-    asset_id: xcm::AssetId,
-  }
+  // The location where the spend is originated (alternative name: `withdraw_context`).
+  spend_context: xcm::Location,
+  // Asset identifier in the context of the `spend_context` location,
+  // equivalent location on chain Z can be derived through re-anchoring.
+  asset_id: xcm::AssetId,
```

(unless you need to carry the `location` of `asset_id` in many places and it makes more sense to keep them wrapped in `LocatableAssetId`

**Comment by muharem** [comment]
@acatangiu renamed (=

But my point is that for cross chain transfers where a transfer command dispatched from X chain, the asset withdraws from Y chain, and deposited on Z chain, where X ≠ Y ≠ Z, you need, the location of asset_id as xcm::Location, the asset_id as xcm::AssetId, the beneficiary location as xcm::Location and beneficiary address as xcm::Location. And this is not specific for the treasury pallet.

I do not understand I you agree with above. 

For the treasury the spend call looks like `spend(origin, asset_kind: T::AssetKind, beneficiary: T::Beneficiary, ...)`. As a client you can setup it to work only locally, or only from local to chain Y. I wish to not leak the xcm details to the frame's treasury pallet.

**Comment by bkontur** [comment]
little bit unrelated, just suggestion, what about changing [`VersionedLocatableAsset`](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/runtime/common/src/impls.rs#L144-L151) - then we don't need to handle/update this `VersionedLocatableAsset` all the time we add/remove XCM versions:
```diff
- pub enum VersionedLocatableAsset {
+ pub struct VersionedLocatableAsset {
-	#[codec(index = 3)]
-	V3 { location: xcm::v3::Location, asset_id: xcm::v3::AssetId },
-	#[codec(index = 4)]
-	V4 { location: xcm::v4::Location, asset_id: xcm::v4::AssetId },
-	#[codec(index = 5)]
-	V5 { location: xcm::v5::Location, asset_id: xcm::v5::AssetId },
+      pub location:  VersionedLocation,
+      pub asset_id:  VersionedAssetId,
}
```

**Comment by muharem** [comment]
@bkontur yes, but lets you to use different versions for location and asset id. may be its fine

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/multi-asset-bounties-now-live-on-westend-testing-and-roadmap/16899/1
